### ISSUE TITLE
Optimisation: do instant query count() without fetching samples

### ISSUE THREAD

**Issue by bboreham** [prometheus/prometheus#10073]
Currently `count()` is implemented as an aggregation, which is called within `rangeEval()` here: https://github.com/prometheus/prometheus/blob/d677aa4b29a8a0cf9f61af04bbf5bfdce893cf23/promql/engine.go#L1191-L1197

`aggregation()` takes as input `[]Sample`, so the storage layer must fetch samples.

In the case it has been called from an instant query, the start and end of the range are equal, but nothing changes, it still fetches samples from storage.

It will be cheaper to compute `count()` the same way we do `/series`, without fetching samples: https://github.com/prometheus/prometheus/blob/c954cd9d1d4e3530be2939d39d8633c38b70913f/tsdb/querier.go#L140-L143

However this requires some refactoring to be able to run the aggregation (`count by (foo,bar)`, etc.) outside of a function taking `[]Sample`.

I suggest doing this only for the instant-query case, because over a time-range series can come and go, so we need per-timestamp information and may as well get that via samples.

**Comment by roidelapluie** [comment]
I am not sure that this optimization is worth it. We have the risk of introducing bugs for a very niche usecase. The complexity seems too high to me.

**Comment by bboreham** [comment]
I disagree this is "very niche"; my guess is every Prometheus admin runs such queries. I edited the description to add that info.

As a data point, on a v2.31.1 server with 2 million series `count({__name__=~".+"})` took 24 seconds; profile says extracting the samples took 8.5 seconds of that.

**Comment by prymitive** [comment]
Running count() is something I do a lot when auditing metrics usage, especially with big instances shared by many teams.
It is expensive, especially counting by `__name__` so I mostly rely on `job` label and scrape metrics provided by Prometheus, but would love if `count({__name__=~".+"})` was cheaper as it would help a lot.

**Comment by Harkishen-Singh** [comment]
Do we plan to implement this? I can try an implementation as a PR.

**Comment by bboreham** [comment]
> we would need to address backwards-compatibility

This came up again; we could have it as a different function name, e.g. `count_for_block`.  
Naming is hard.

**Comment by beorn7** [comment]
Hello from the bug scrub!

Since this came up again as recently as February (see above), we'll leave it open. However, it feels weird to introduce a new function just for a specific performance optimization. Maybe this should be an API call outside of PromQL?

(This reminds me of my wild meta data ideas, which might include information about start and end point of series, so that we could know about the end of a series without decoding samples to find staleness markers… With that in place, we could optimize in this way without breaking semantics.)
